endpoint: create distinct type for restoring

ianvernon opened the pull request:
The Endpoint structure has been tied to contain specific fields which specific
names, as well as a flat structure due to the need to have compatibility upon
upgrades from one version of Cilium to the next. However, this means that
doing refactoring of the internals of the Endpoint structure is impossible,
as it has to maintain these API guarantees. To allow for more flexibility in
changing the structure of the Endpoint, instead, create another type which
clearly specifies what fields in the Endpoint are needed to be serialized into
its headerfile, which is then read upon restore after cilium-agent restarts.
The forward-compatible guarantee is now placed on this type instead of the
Endpoint itself, allowing for future refactoring of Endpoint without need to
worry about breaking API guarantees.

For now, do not make any changes to the Endpoint type itself to avoid coupling
refactoring changes with this PR.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8941)
<!-- Reviewable:end -->

Commit:
endpoint: create distinct type for restoring

The Endpoint structure has been tied to contain specific fields which specific
names, as well as a flat structure due to the need to have compatibility upon
upgrades from one version of Cilium to the next. However, this means that
doing refactoring of the internals of the Endpoint structure is impossible,
as it has to maintain these API guarantees. To allow for more flexibility in
changing the structure of the Endpoint, instead, create another type which
clearly specifies what fields in the Endpoint are needed to be serialized into
its headerfile, which is then read upon restore after cilium-agent restarts.
The forward-compatible guarantee is now placed on this type instead of the
Endpoint itself, allowing for future refactoring of Endpoint without need to
worry about breaking API guarantees.

For now, do not make any changes to the Endpoint type itself to avoid coupling
refactoring changes with this PR.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani commented on pkg/endpoint/restore.go:
```diff
@@ -88,3 +95,135 @@ func ReadEPsFromDirNames(owner regeneration.Owner, basePath string, eptsDirNames
+	DatapathConfiguration models.EndpointDatapathConfiguration
+}
+
+func (r *restoredEndpoint) populateEndpoint(ep *Endpoint) {
```
What was the rationale to make this a method on restoredEndpoint? I would have expected a Endpoint.fromRestoredEndpoint (or a constructor but this is already being used in that function so it's more of a utility method).

ianvernon replied:
Switched the order, your reasoning makes sense.

raybejjani commented on pkg/endpoint/restore.go:
```diff
@@ -88,3 +95,135 @@ func ReadEPsFromDirNames(owner regeneration.Owner, basePath string, eptsDirNames
+// marked as private to JSON marshal. Do NOT modify this structure in ways which
+// is not JSON forward compatible.
+//
+type restoredEndpoint struct {
```
`serializableEndpoint` might better capture the intent here. `restoredEndpoint` might imply that it's already restored rather than the endpoint being restored. Not a big deal either way.

ianvernon replied:
Done!

raybejjani commented on pkg/endpoint/endpoint.go:
```diff
@@ -109,15 +109,6 @@ var _ notifications.RegenNotificationInfo = &Endpoint{}
-// will read the file and re-create all endpoints with all fields which are not
-// marked as private to JSON marshal. Do NOT modify this structure in ways which
-// is not JSON forward compatible.
-//
```
It may be good to leave a note here saying that the serializable bit is now in `restoredEndpoint`. This is mostly for backports (which I'm hoping won't apply cleanly once we merge in other changes to Endpoint) so they know where to look.

ianvernon replied:
Done!

raybejjani reviewed (changes requested):
I'm fine with adding these methods as-is if we have a reason for them later, but could we instead implement https://golang.org/pkg/encoding/json/#Marshaler and https://golang.org/pkg/encoding/json/#Unmarshaler ? This might make some of the logs we add here more difficult to have, and I do like having them. It also means that `restoredEndpoint` is a simple data holder.

Commit:
endpoint: rename `restoredEndpoint` to `serializableEndpoint`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: clarify relation of `serializableEndpoint` to `Endpoint`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: swap receiver / parameters for populating Endpoint from serializableEndpoint

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: implement `MarshalJSON` and `UnmarshalJSON`

This hides that the endpoint is converted intermediately to a
`serializableEndpoint`, and implements the `Marshaler` and `Unmarshaler`
interfaces.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani reviewed (approved):
Very LGTM. Thanks :)
